include/nbl/builtin/glsl/ext/Scan/scan.glsl

```diff
@@ -0,0 +1,122 @@
+#ifndef _NBL_GLSL_EXT_SCAN_INCLUDED_
+#define _NBL_GLSL_EXT_SCAN_INCLUDED_
+
+#include <nbl/builtin/glsl/workgroup/arithmetic.glsl>
+#include <nbl/builtin/glsl/math/typeless_arithmetic.glsl>
+#include <nbl/builtin/glsl/limits/numeric.glsl>
+
+#include "nbl/builtin/glsl/ext/Scan/parameters.glsl"
+
+#ifndef _NBL_GLSL_EXT_SCAN_GET_PADDED_DATA_DECLARED_
+
+uint nbl_glsl_ext_Scan_getPaddedData(in uint idx, in uint pad_val, bool is_upsweep);
+int nbl_glsl_ext_Scan_getPaddedData(in uint idx, in int pad_val, bool is_upsweep);
+float nbl_glsl_ext_Scan_getPaddedData(in uint idx, in float pad_val, bool is_upsweep);
+
+#define _NBL_GLSL_EXT_SCAN_GET_PADDED_DATA_DECLARED_
+#endif
+
+#ifndef _NBL_GLSL_EXT_SCAN_GET_PARAMETERS_DEFINED_
+#error "You need to define `nbl_glsl_ext_Scan_getParameters` and mark `_NBL_GLSL_EXT_SCAN_GET_PARAMETERS_DEFINED_`!"
+#endif
+
+#ifndef _NBL_GLSL_EXT_SCAN_GET_PADDED_DATA_DEFINED_
+#error "You need to define `nbl_glsl_ext_Scan_getPaddedData` and mark `_NBL_GLSL_EXT_SCAN_GET_PADDED_DATA_DEFINED_`!"
+#endif
+
+//
+// Upsweep
+//
+
+#define NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(BIN_OP_NAME, TYPE, IDENTITY)\
+void nbl_glsl_ext_Scan_upsweep##BIN_OP_NAME(in uint idx, out TYPE val)\
+{\
+	TYPE data = nbl_glsl_ext_Scan_getPaddedData(idx, IDENTITY, true);\
+	val = (gl_NumWorkGroups.x == 1u) ? nbl_glsl_workgroupExclusive##BIN_OP_NAME(data) : nbl_glsl_workgroupInclusive##BIN_OP_NAME(data);\
```

Comment:
use a proper if-statement for this, ternary op kinda suggests a cmov (not that compilers honour it).